pkg/machines/libvirt-dbus.es6

```diff
@@ -0,0 +1,861 @@
+                "Defined": 0,
+                "Undefined": 1,
+                "Started": 2,
+                "Suspended": 3,
+                "Resumed": 4,
+                "Stopped": 5,
+                "Shutdown": 6,
+                "PMsuspended": 7,
+                "Crashed": 8
+            };
+            let objPath = args[0];
+            let eventType = args[1];
+
+            logDebug(`signal on ${path}: ${iface}.${signal}(${JSON.stringify(args)})`);
+
+            switch (eventType) {
+            case domainEvent["Defined"]:
+                dispatch(getVm({connectionName, id:objPath}));
+                break;
+
+            case domainEvent["Undefined"]:
+                dispatch(undefineVm({connectionName, id: objPath}));
+                break;
+
+            case domainEvent["Started"]:
+                dispatch(getVm({connectionName, id:objPath}));
+                break;
+
+            case domainEvent["Suspended"]:
+                dispatch(updateVm({
+                    connectionName,
+                    id: objPath,
+                    state: 'paused'
+                }));
+                break;
+
+            case domainEvent["Resumed"]:
+                dispatch(updateVm({
+                    connectionName,
+                    id: objPath,
+                    state: 'running'
+                }));
+                break;
+
+            case domainEvent["Stopped"]:
+                dispatch(getVm({
+                    connectionName,
+                    id: objPath
+                }));
+                // transient VMs don't have a separate Undefined event, so remove them on stop
+                dispatch(undefineVm({connectionName, id: objPath, transientOnly: true}));
+                break;
+
+            default:
+                logDebug(`Unhandled lifecycle event type ${eventType}`);
+                break;
+            }
+        }
+    );
+
+    /* Subscribe to signals on Domain Interface */
+    clientLibvirt[connectionName].subscribe(
+        { interface: 'org.libvirt.Domain' },
+        (path, iface, signal, args) => {
+            logDebug(`signal on ${path}: ${iface}.${signal}(${JSON.stringify(args)})`);
+
+            switch (signal) {
+            case 'ControlError':
+            case 'DeviceAdded':
+            case 'DeviceRemoved':
+            case 'DiskChange':
+            case 'MetadataChanged':
+            case 'TrayChange':
+            /* These signals imply possible changes in what we display, so re-read the state */
+                dispatch(getVm({connectionName, id:path}));
+                break;
+
+            default:
+                logDebug(`handleEvent ${connectionName} ${name}: ignoring event ${signal}`);
+            }
+        });
+
+    /* Check stopped libvirtd */
+    let systemdClient = cockpit.dbus('org.freedesktop.systemd1', { bus: connectionName });
+    systemdClient.subscribe(
+        { interface: 'org.freedesktop.systemd1.Manager', member: 'JobRemoved' },
```

Comment:
I'm ok with this, but rather than introducing assumptions and knowledge about systemd here, it seems to me it would be more elegant and straightforward to monitor when the `clientLibvirt[connectionName]` object goes away. You can use the [track option of cockpit.dbus()](https://cockpit-project.org/guide/latest/cockpit-dbus.html) for that and subscribe to the "close" signal.

This will then work in the same way for the user libvirt instance, which is (most likely) not started by systemd, but on demand by the library or the session bus.

Reply:
I tried the following but the hander gets never triggered:
```
-    let systemdClient = cockpit.dbus('org.freedesktop.systemd1', { bus: connectionName });
-    systemdClient.subscribe(
-        { interface: 'org.freedesktop.systemd1.Manager', member: 'JobRemoved' },
-        (path, iface, signal, args) => {
-            logDebug(`signal on ${path}: ${iface}.${signal}(${JSON.stringify(args)})`);
-
-            if (args[2] == 'libvirtd.service') {
-                dispatch(checkLibvirtStatus(libvirtServiceName));
-                dispatch(deleteUnlistedVMs(connectionName, []));
-                dispatch(delayPolling(getAllVms(connectionName, libvirtServiceName)));
-            }
-        });
+    cockpit.dbus('org.libvirt', { bus: connectionName, "track": true }).addEventListener(
+        "close",
+        () => {
+            logDebug("Libvirt went away from D-Bus");
+
+            dispatch(checkLibvirtStatus(libvirtServiceName));
+            dispatch(deleteUnlistedVMs(connectionName, []));
+            dispatch(delayPolling(getAllVms(connectionName, libvirtServiceName)));
+        }
+    );
```

Did you mean something different?
I tried to trigger it by restarting libvirtd or kill the libvirt-dbus process but this handler doesn't execute.

Reply:
Did you literally use this? That's not the cockpit.dbus() client you are really using, but an entirely new instance, isn't it? I suppose that just immediately gets GCed away again. Can you try on `clientLibvirt[connectionName]`?

We do that exact thing in the [packagekit lib](https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/packagekit.es6#L70), but on a long-lived global variable, where it works just fine (we have a test for that).

Reply:
So indeed it works, I just misunderstood what the signal is doing. So, it the existing code I am getting the signal when libvirtd is not running, thus I am able to run the checkLibvirtStatus and display a page showing 'Libvirtd service is not active'...
When I am listening to the libvirt-dbus close signal, I just get triggered when libvirt-dbus process existed which is not attached to libvirtd. libvirtd can be stopped, but the libvirt-dbus process running, so I will not get notified for a killed libvirtd. Does this make any sense?
Do you have any other idea how to track stopped libvirtd apart from the current implementation?
Probably I should also keep the EventListener to close signal on dbus libvirt. since they serve slightly different purpose. 

Note: I can't change the behavior of tracking libvirtd status changes because I bind to have same behavior are libvirt-virsh.js implementation.

Reply:
Oh right, I didn't consider that libvirt.service != libvirt-dbus. So I suppose this is okay for now. But how do you track the user instance then?

Reply:
I don't track it actually. I couldn't think of some way to track it. Any suggestions?

Reply:
BTW I added the EventListener to close signal for libvirt-dbus as well. I think it makes sence to have both libvirtd and libvirt-dbus close handling.

Reply:
I'm not sure whether it make sense to track libvirt-dbus.  Since it's a dbus daemon it gets started on demand and if it crashes it will be started again once we do some dbus call.  The only way how it can disappear is that it's removed from system completely.

We don't need to track the session (user) instance as well, the libvirtd for session instance is again started on demand, if it crashes/is killed it will be started again unless it's removed from system completely.

Reply:
OK, good to know. I was about to suggest that libvirt-dbus should notice when libvirt goes down, and die with it -- then libvirt can transitively be tracked through the D-Bus "close" signal. But it seems that this isn't actually necessary. Let's not dwell on this then -- if there is some fine-tuning to be made, it can become a follow-up.

Reply:
Just for the review ease, I removed EventListener to close signal according to this discussion.